lib/src/main/kotlin/com/swmansion/starknet/data/ContractAddressCalculator.kt

```diff
@@ -53,4 +57,40 @@ object ContractAddressCalculator {
         salt = salt,
         deployerAddress = Felt.ZERO,
     )
+
+    /**
+     * Checks if provided string is in a checksum address format.
+     *
+     * @param address to check
+     */
+    @JvmStatic
+    fun isChecksumAddressValid(address: String): Boolean {
+        return calculateChecksumAddress(Felt.fromHex(address)) == address
+    }
+
+    /**
+     * Outputs formatted checksum address.
+     *
+     * Follows implementation of starknet.js. EIP55 was an inspiration, but it is not compatible as it treats hex string
+     * as encoded number, instead of encoding it as ASCII string.
+     * Example checksum address - 0x02Fd23d9182193775423497fc0c472E156C57C69E4089A1967fb288A2d84e914.
+     *
+     * @param address to encode
+     */
+    @JvmStatic
+    fun calculateChecksumAddress(address: Felt): String {
+        val stringAddress = address.value.toString(16).lowercase(Locale.ENGLISH).padStart(64, '0')
+        val chars = stringAddress.toCharArray()
+        // This part is different from EIP55. EIP55 digests hex as encoded ASCII string and here we
+        // use digest it as bytes.
+        val hashed = keccak(BigInteger(stringAddress, 16).toBytes())
+        for (i in chars.indices) {
+            // We subtract from 256, because we use bits starting from the end
+            if (chars[i].isLetter() && hashed.testBit(256 - 4 * i - 1)) {
```

Comment:
I thought about it and just the logic for reverting it would be more complex. An alternative would be to process it as strings, but I like that even less as it is clunky - we can ask for specific bits right away, so turing it into hex and then parsing doesn't make sense.
Very often low level logic is not obvious unfortunately.

Reply:
Okay, let's hope whoever will have to modify it in the future will understand what's going on here 😅